### ISSUE TITLE
cooperator should provide an API to pause, resume, and cancel individual tasks within the cooperator instead of the entire thing.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2712]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#2712|
|Type|enhancement|
|Created|2007-06-22 18:56:25Z|
|Branch|https://github.com/twisted/twisted/tree/coop-2712-5|

The need for this feature arose out of some requirements that [Etsy](http://www.etsy.com/) discovered in the use of Twisted as a scheduler for several long-running batch tasks.  They are providing some help with the implementation.

This implies the need for a new Cooperator API which does not simply return a ```Deferred```.

<details><summary>Searchable metadata</summary>

```
trac-id__2712 2712
type__enhancement enhancement
reporter__glyph glyph
priority__high high
milestone__ 
branch__branches_coop_2712_5 branches/coop-2712-5
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1182538585000000 1182538585000000
changetime__1244705079000000 1244705079000000
version__None None
owner__glyph glyph
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph removed owner|
|-|-|

I could have sworn I wanted to add something else to this, but now that I'm looking at it it seems fine.  It's ready for review in ```coop-2712```.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

That looks really nice. I have only a few comments:
 * the test-case-name of task.py should be ```twisted.test.test_task,twisted.test.test_cooperator``` instead of just ```twisted.test.test_cooperator```
 * test_cooperator needs a copyright notice, maybe a docstring too.
 * PauseResumeTests.worker is missing a docstring.
 * there is a _addTask method on Cooperator, but you use directly _tasks.remove. Maybe a _removeTask would make it more symmetric.

That's it!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @glyph|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

(In [20764]) re [#2712](https://github.com/twisted/twisted/issues/2712) deal with review feedback

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph removed owner|
|-|-|

With the revision conveniently hyperlinked (if you arrive late enough for trac synchronization) I think I've dealt with everything.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to @glyph|
|-|-|

Last nitpicking:
 * in ```PauseResumeTests.worker```, you use 'while 1', I'd prefer 'while True'
 * in test_pauseWhileDeferred docstring, you use 'callsback': I guess 'callbacks' or 'calls back' would be better. Furthermore, 'the pause is resumed' doesn't make really sense to me (but that could be me :))
 * the docstring of test_cooperator could have a L{} to twisted.internet.task.Cooperator

Then please merge!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

The convention thus far has not been to document parameters in ```__init__``` which are only set as attributes on the object in ```__init__```'s docstring, but instead to document them in the class docstring.  In any case, ```CooperativeTask``` is missing documentation for ```pauseCount```.

There's no test coverage for the ```doneDeferred``` parameter to the new ```cooperate``` method.  I'd suggest dropping it.

```CooperativeTask.getDeferred``` seems like it could be named more usefully, or dropped.  Cooperator uses it in one place, and ```CooperativeTask.deferred``` in five places.  The tests don't exercise it directly at all.

It also seems like ```CooperativeTask``` would be the right place to put most of the logic in _tick.  There's no functional implication of such a refactoring though, so I won't suggest it's necessary in order for this branch to be merged.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph removed owner|
|-|-|

Replying to [therve](#note_7):

  * ''in ```PauseResumeTests.worker```, you use 'while 1', I'd prefer 'while True' ''
    * But Python's bytecode compiler can recognize and optimize the 'while 1' loop!!!  (Okay, seriously, done.)
  * _in test_pauseWhileDeferred docstring, you use 'callsback': I guess 'callbacks' or 'calls back' would be better. Furthermore, 'the pause is resumed' doesn't make really sense to me (but that could be me :))_
    * Looking at that docstring it was a complete mess.  I've fixed it to read completely differently.

Replying to [exarkun](#note_8):

  * ''The convention thus far has not been to document parameters in ```__init__``` which are only set as attributes on the object in ```__init__```'s docstring, but instead to document them in the class docstring. In any case, ```CooperativeTask``` is missing documentation for pauseCount.''
    * Thanks for this feedback; it brought to my attention the fact that none of those attributes are public.  They're all documented, but private, now.
  * _There's no test coverage for the doneDeferred parameter to the new cooperate method. I'd suggest dropping it._
    * Good point.  There was some duplication before when I tried to do that, but I realized it was somewhat pointless and removed it.
  * ''```CooperativeTask.getDeferred``` seems like it could be named more usefully, or dropped. Cooperator uses it in one place, and ```CooperativeTask.deferred``` in five places. The tests don't exercise it directly at all.''
    * Removed.
  * ''It also seems like ```CooperativeTask``` would be the right place to put most of the logic in _tick. There's no functional implication of such a refactoring though, so I won't suggest it's necessary in order for this branch to be merged.''
    * I like the idea that one object or the other is completely responsible for calls to _addTask/_removeTask, so I went ahead and did it.  Then I discovered I could remove the gross capture-the-loop-variable keyword args trick and was ''very'' glad I did.

The code is still in ```coop-2712``` and ready for what will hopefully be its final review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set owner to @glyph|
|-|-|

 * doneDeferred is still in the docstring of cooperate
 * a toplevel cooperate function using ```_theCooperator``` would probably be handy
 * you've removed the getDerred method, and change the deferred attribute to be private: I think that means there's no way to get the end of the task when using cooperate without accessing the private attribute
 * typo in ```_oneWorkUnit```: retrieiving
 *  in the docstring of ```CooperativeTask```, you say that the task can be stopped, but this can't be done without stopping the cooperator completely. The API is maybe missing a stop method.
 * the ```_iterator``` docstring use the 'I am' style, which we try to avoid :)
 * As we removed the doneDeferred of cooperate, we should probably deprecate it for coiterate, unless I missed something (to be done in another ticker)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

(In [22476]) Branching to 'coop-2712-2'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph commented|
|-|-|

(In [22479]) re [#2712](https://github.com/twisted/twisted/issues/2712) this sucks but so do all singletons.  at some point the rubber meets the road, I guess.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph set owner to @therve<br>@glyph set status to `new`|
|-|-|

Replying to [therve](#note_10):

I think I've done with all the feedback except these...

  * _you've removed the getDeferred method, and change the deferred attribute to be private: I think that means there's no way to get the end of the task when using cooperate without accessing the private attribute_ (and)
  * _As we removed the doneDeferred of cooperate, we should probably deprecate it for coiterate, unless I missed something (to be done in another ticker)_
    * I think that a completely different API is actually required here.  There's a state-sharing problem.  If you're calling an API and getting a Deferred back, you can generally safely assume that the Deferred is "yours", and you can add a callback to it and get a sensible result.  In this case, getDeferred (as it was implemented) provides an API where you get a Deferred which you don't know where it's been before.  This means you don't know what callbacks and errbacks to expect, and if you modify it in any way, the next caller might not get the right thing.  I wrote a [http://divmod.org/trac/browser/trunk/Epsilon/epsilon/pending.py succinct, but not particularly good solution] to this for Divmod at one point, which has only been lightly used.  Perhaps Twisted should have something like this.  More specifically I think that we should have a separate ticket for getting a completion Deferred from a `CooperativeTask`, and Twiste dmight grow a `PendingEvent` (but gooder) to implement it.  This same ticket can deprecate `doneDeferred` and cooperate entirely since it will be a complete replacement.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve set owner to @glyph|
|-|-|

 * _completed and _failed attributes are not documented (but look at the next comments first :))
 * the _checkStopped docstring is a bit outdated, and thus the one of stop too.
 * I wonder if the 3 flags _stopped/_completed/_failed could not be replaced a simple state variable
 * `AliasTests.test_coiterate` should probably be named test_cooperate
 * there is a trailing whitespace in test_cooperator
 * the indentation of the comments in test_stopDeferred is a bit fancy

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_17"></a>|@therve commented|
|-|-|

Hey glyph don't let this rotten!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph commented|
|-|-|

(In [26526]) Branching to 'coop-2712-3'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph commented|
|-|-|

(In [26528]) re [#2712](https://github.com/twisted/twisted/issues/2712), address these comments:
    * ''there is a trailing whitespace in test_cooperator''
    * ''the indentation of the comments in test_stopDeferred is a bit fancy''

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph commented|
|-|-|

(In [26529]) re [#2712](https://github.com/twisted/twisted/issues/2712), address this comment:

    * AliasTests.test_coiterate should probably be named test_cooperate

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph commented|
|-|-|

(In [26530]) re [#2712](https://github.com/twisted/twisted/issues/2712), address these comments:

    * ''the _checkStopped docstring is a bit outdated, and thus the one of stop too.''
    * ''I wonder if the 3 flags _stopped/_completed/_failed could not be replaced a simple state variable''
    * ''_completed and _failed attributes are not documented (but look at the next comments first :))''

I've replaced the various flags by creating the appropriate exception and assigning that to the state variable, which is None if it's not done yet.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_22"></a>|@glyph commented|
|-|-|

Build results here: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/coop-2712-3

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_23"></a>|@glyph removed owner|
|-|-|

Most builds look good at this point; let me know what you think.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_24"></a>|@therve set owner to @glyph|
|-|-|

```
+    def test_pauseWhileDeferred(self):
+        """
+        C{pause()}ing a task is while it is waiting on an outstanding
```

It seems there is an extra word here.

```
+        @param doneDeferred: If specified, this will be the Deferred used as
+        the completion deferred.  It is suggested that you use the default,
+        which creates a new Deferred for you.
```

The second part under `@param` should be indented.

Please also update copyright notices of both files to 2009.

This looks good I think, but I'm wondering a thing about the API: aren't we missing a way to get the `_deferred` attribute of the task when we call `cooperate`? There has been lots of back and forth on this particular subject, so maybe I'm missing something.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_25"></a>|@glyph commented|
|-|-|

(In [26537]) re [#2712](https://github.com/twisted/twisted/issues/2712) Update copyright statements.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_26"></a>|@glyph commented|
|-|-|

(In [26538]) re [#2712](https://github.com/twisted/twisted/issues/2712) Address extra word in docstring.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_27"></a>|@glyph commented|
|-|-|

(In [26539]) re [#2712](https://github.com/twisted/twisted/issues/2712) Indent after @param.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_28"></a>|@glyph commented|
|-|-|

You're right.  We are missing a way to get at the `_deferred` attribute.  Accessing it directly is bad, because you can't add a callback with any expectations of its return value.  So, we should have something that generates a new `Deferred` for the caller for each listener.

There are some other issues that occurred to me after re-reading this code too; for example, the private access to `_deferred` in `dieNext` is problematic; it means that there will be no reasonable way to quash errors from being logged.  So I'm going to sleep on it.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_30"></a>|@glyph commented|
|-|-|

(In [26963]) Branching to 'coop-2712-4'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_31"></a>|@glyph removed owner<br>@glyph set status to `new`|
|-|-|

Here's a different strategy, one I think is probably quite a bit better.

[Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/coop-2712-4)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_32"></a>|@therve set owner to @glyph|
|-|-|

OK! I've been looking at it for a while, and I think I'm happy with it. Please merge!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_33"></a>|@glyph set status to `closed`|
|-|-|

Gah.  Jumped the gun on the commit, so I didn't get the 'fixes' into the message.  This was closed as of r26982.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_34"></a>|@glyph set status to `reopened`|
|-|-|

(In [26989]) Revert r26982 - Reopens [#2712](https://github.com/twisted/twisted/issues/2712) - due to an incompatibility introduced into .stop() that causes failures in the Axiom test suite.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_35"></a>|@glyph commented|
|-|-|

(In [26991]) Branching to 'coop-2712-5'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_36"></a>|@glyph commented|
|-|-|

(In [26993]) re [#2712](https://github.com/twisted/twisted/issues/2712) add a test for reentrancy

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_37"></a>|@glyph commented|
|-|-|

(In [26994]) re [#2712](https://github.com/twisted/twisted/issues/2712) move the task-removal to an earlier phase to prevent reentrancy

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_38"></a>|@glyph removed owner<br>@glyph set status to `new`|
|-|-|

Reentrancy issues fixed; I've added a test and axiom's suite passes now.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_40"></a>|@jml set owner to @glyph|
|-|-|

Hey Glyph,

I'm assuming that exarkun & therve have reviewed everything up to 26991, and so I don't have to check the whole branch again.

  * The change itself -- calling `_removeTask` before the callbacks -- is fairly subtle. Could you please add a comment explaining why the ordering is important? See twisted/internet/task.py:338.

  * The test is also subtle (yay concurrency!). It's not clear to me what value the `otherwork` stuff adds to the test -- the assertion passes with or without the change. Can you please explain this to me before you land the branch?

Thanks for figuring this out & adding the test. I'll be online for another hour or so.

jml

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_41"></a>|@glyph commented|
|-|-|

(In [26999]) re [#2712](https://github.com/twisted/twisted/issues/2712) As jml pointed out, the extraneous generator doesn't really add anything to the test besides confusion

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_42"></a>|@glyph commented|
|-|-|

(In [27000]) re [#2712](https://github.com/twisted/twisted/issues/2712) Try to explain in a comment.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @jml|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_44"></a>|@jml set owner to @glyph|
|-|-|

Looks good to me. Thanks for clearing that up.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_45"></a>|@glyph set status to `closed`|
|-|-|

(In [27001]) Add an API to pause, resume, and cancel individual tasks within a `Cooperator`.

This also adds an API to provide Deferreds to multiple consumers from an individual task.

The new API is in `twisted.internet.task.cooperate` and `twisted.internet.task.CooperativeTask`.

Author: glyph

Reviewer: therve, jml

Fixes [#2712](https://github.com/twisted/twisted/issues/2712)
